WebApplication1/WebApplication1/Views/ToDoLists/Index.cshtml

```diff
@@ -3,46 +3,156 @@
+                        </tr>
+                    </tbody>
+                    <tfoot>
+                        <tr>
+                            <td colspan="2">
+                                <button data-bind="click: addItem">Add</button>
+                            </td>
+                        </tr>
+                    </tfoot>
+                </table>
+            </div>
+        </div>
+    </div>
+    <!-- /ko -->
+</div>
 @section Scripts
-    {
+{
     <script type="text/javascript">
-        /*var vm = {
-            hello: function() {
-                alert("hello");
+        function TodoItem(host, id) {
+            this.host = host;
+            this.todoListId = host.id;
+            this.id = id;
+            this.isEditMode = ko.observable(false);
+            this.toggleEditMode = function() {
+                this.isEditMode(!this.isEditMode());
+            };
+            this.memo = ko.observable();
+            this.isDone = ko.observable();
+            var self = this;
+            this.isDone.subscribe(function(value) {
+                self.isDone(value);
+                self.host.postData(self);
+            });
+            this.memo.subscribe(function(value) {
+                self.memo(value);
+                self.host.postData(self);
+                self.editModeOff();
+            });
+            this.editModeOff = function() {
+                self.isEditMode(false);
+            };
+
+        }
+
+        function TodoList(host, id) {
+            this.host = host;
+            this.id = id;
+            this.name = ko.observable();
+            this.items = ko.observableArray([]);
+            this.addItem = function() {
+                this.items.push(new TodoItem(this, 0));
+            };
+            this.postData = function(item) {
+                if (this.host.ignore)
+                    return;
+
+                var self = this;
+                $.ajax({
+                    type: "post",
+                    dataType: "JSON",
+                    url: '@Url.Action("Save", "TodoItems")',
+                    data: {
+                        id: item.id,
+                        todoListId: self.id,
+                        memo: item.memo(),
+                        isDone: item.isDone()
+                    },
+                    success: function(response) {
+                        self.id = response;
+                    }
+                });
+            };
+
+        }
+
+        function ViewModel() {
+            var self = this;
+            this.ignore = true;
+            this.lists = ko.observableArray([]);
+            this.addList = function() {
+                var name = prompt("Enter a name for the list: ");
+                if (name.length === 0)
+                    return;
+
+                var newList = new TodoList(this, 0);
+                newList.name(name);
+                this.postData(newList);
+                this.lists.push(newList);
```

Comment:
Consider re-introducing the `ViewModel.saved()` method and calling it from the `$.ajax` configuration's `success` hook in the `ViewModel.postData()` method. This would allow us to easily manage the `this.lists()` variable in one place. Something like this:

```javascript
function ViewModel() {

  this.self = this;

  // other code omitted for brevity...

  this.postData = function(list) {
    $.ajax({
      ...
      success: function(response) {
        self.saved(response, list);
      }
    });
  };

  this.saved = function(response, list) {
    if (list.id <= 0) {
      list.id = response;
      self.lists.push(list);
    }
    // more code could go here...
    console.log("Saved!");
    list.items.push(new TodoItem("", 0)); // let user add new item right away
  };

}
```